File: yfood/src/main/kotlin/com/jornadadev/yfood/pagamentoonline/gateway/GatewaySaori.java
package com.jornadadev.yfood.pagamentoonline.gateway;

import com.jornadadev.yfood.entities.Pagamento;
import com.jornadadev.yfood.entities.Transacao;
import com.jornadadev.yfood.pagamentoonline.RequestsGateways;
import com.jornadadev.yfood.pagamentoonline.Resultado;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import javax.validation.Valid;
import java.math.BigDecimal;

@Service
public class GatewaySaori extends Gateway {

//	private String id = "gateway-saori";
//
//	private RequestsGateways requestsGateways;
//
//	public GatewaySaori(RequestsGateways requestsGateways) {
//		super();
//		this.requestsGateways = requestsGateways;
//	}
//
//	@Override
//	public boolean aceiteEspecifico(
//			@NotNull @Valid Pagamento novoPagamentoSalvo) {
//		FormaPagamento formaPagamento = novoPagamentoSalvo.getFormaPagamento();
//		return formaPagamento.pertence(FormaPagamento.visa,
//				FormaPagamento.master);
//	}
//
//	@Override
//	public Resultado<Exception, Transacao> processaEspecifico(
//			@NotNull @Valid Pagamento pagamento) {
//
//		log.debug("Processando pagamento por gateway Saori");
//		DadosCompraGenerico request = new DadosCompraGenerico(pagamento);
//		requestsGateways.saoriProcessa(request);
//		return Resultado.sucesso(Transacao.concluida(this));
//	}
//
//	@Override
//	public BigDecimal custoEspecifico(Pagamento pagamento) {
//		return pagamento.getValor().multiply(new BigDecimal("0.05"));
//	}
//
//	@Override
//	public String toString() {
//		return "GatewaySaori []";
//	}
//
//	@Override
//	public int hashCode() {
//		final int prime = 31;
//		int result = 1;
//		result = prime * result + ((id == null) ? 0 : id.hashCode());
//		return result;
//	}
//
//	@Override
//	public boolean equals(Object obj) {
//		if (this == obj)
//			return true;
//		if (obj == null)
//			return false;
//		if (getClass() != obj.getClass())
//			return false;
//		GatewaySaori other = (GatewaySaori) obj;
//		if (id == null) {
//			if (other.id != null)
//				return false;
//		} else if (!id.equals(other.id))
//			return false;
//		return true;
//	}

}

File: yfood/src/main/kotlin/com/jornadadev/yfood/pagamentoonline/gateway/GatewaySaori.kt
package com.jornadadev.yfood.pagamentoonline.gateway

import com.jornadadev.yfood.entities.FormaPagamentoEnum
import com.jornadadev.yfood.entities.Pagamento
import com.jornadadev.yfood.entities.Transacao
import com.jornadadev.yfood.outrossistemas.DadosCompraGenerico
import com.jornadadev.yfood.pagamentoonline.RequestsGateways
import com.jornadadev.yfood.pagamentoonline.Resultado
import org.springframework.stereotype.Service
import java.math.BigDecimal
import javax.validation.Valid
import javax.validation.constraints.NotNull

@Service
class GatewaySaori(private val requestsGateways: RequestsGateways) : Gateway() {

    private val id: String = "gateway-saori"

    public override fun aceiteEspecifico(novoPagamentoSalvo: @Valid Pagamento): Boolean {
        val formaPagamento = novoPagamentoSalvo.formaPagamento
        return formaPagamento.pertence(FormaPagamentoEnum.VISA,
                FormaPagamentoEnum.MASTERCARD)
    }

    override fun processaEspecifico(
            pagamento: Pagamento): Resultado<Exception, Transacao> {
        println("Processando pagamento por gateway Saori")
        val request = DadosCompraGenerico(pagamento)
        requestsGateways.saoriProcessa(request)
        return Resultado.sucesso(Transacao.concluida(this))
    }

    override fun custoEspecifico(pagamento: Pagamento): BigDecimal {
        return pagamento.valor.multiply(BigDecimal("0.05"))
    }

    override fun toString(): String {
        return "GatewaySaori []"
    }

    override fun hashCode(): Int {
        val prime = 31
        var result = 1
        result = prime * result + (id?.hashCode() ?: 0)
        return result
    }

    override fun equals(obj: Any?): Boolean {
        if (this === obj) return true
        if (obj == null) return false
        if (javaClass != obj.javaClass) return false
        val other = obj as GatewaySaori
        if (id == null) {
            if (other.id != null) return false
        } else if (id != other.id) return false
        return true
    }
}